### ISSUE TITLE
Guidance on Improving Contrastive Learning with Additional Features and Fine-tuning on Custom Dataset

### ISSUE THREAD

**Issue by aswiniitkgp** [PengYueECUST1997/CLEF#3]
Hello,

Thank you for sharing this implementation, it has been very helpful in understanding the approach used in the paper. I am currently working on extending this model and would appreciate your guidance on a few points:

- Enhancing Contrastive Learning:

  I would like to improve the contrastive learning framework by incorporating an additional set of features. Are there any 
  recommended or simplified steps to achieve this within the current architecture?
  

- Model Fine-tuning:

  I am also looking to fine-tune the model using my own dataset. Could you please advise on the best practices or steps to adapt the 
  model for this purpose?

Any suggestions or pointers to relevant parts of the code would be greatly appreciated.

Thank you...

Best regards,
Aswin

**Comment by YihangLin123** [comment]
Hello, I would like to ask how to obtain indicators for evaluating model performance and how to test model performance. I look forward to your reply. Thank you.

**Comment by PengYueECUST1997** [comment]
Dear linyihang1215,
 
Thank you very much for your kind message and interest in our work.
 
Regarding your question on how to evaluate the model performance on an independent test set: since our task is a binary classification problem, the calculation of performance metrics such as Accuracy (ACC), F1-score, and Matthews Correlation Coefficient (MCC) is relatively straightforward.
 
Once you have followed the instructions in the repository to perform predictions on your test set, you should obtain a CSV file containing the sample names and the corresponding predicted labels (‘No’ or ‘Yes’). For the provided test set, the true labels can be inferred based on the sample IDs: for example, when predicting T3SEs, samples whose names contain the “T3SE” tag are considered positive (label = 1), and others are negative (label = 0).
 
After obtaining both the true label array and the predicted label array, you can easily compute the performance metrics using the functions provided by scikit-learn. Here’s a simple Python example:
 
```python
 
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, matthews_corrcoef
 
&nbsp;
 
# Assume y_true and y_pred are the true and predicted label arrays
 
acc = accuracy_score(y_true, y_pred)
 
precision = precision_score(y_true, y_pred)
 
recall = recall_score(y_true, y_pred)
 
f1 = f1_score(y_true, y_pred)
 
mcc = matthews_corrcoef(y_true, y_pred)
 
&nbsp;
 
print(f"Accuracy: {acc:.4f}")
 
print(f"Precision: {precision:.4f}")
 
print(f"Recall: {recall:.4f}")
 
print(f"F1-score: {f1:.4f}")
 
print(f"MCC: {mcc:.4f}")
 
```
 
Thank you again for your message — I hope this reply helps you.
 
Best regards,
 Yue Peng



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PengYueECUST1997/CLEF"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年6月6日(星期五) 上午10:31
***@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;Re: [PengYueECUST1997/CLEF] Guidance on Improving Contrastive Learning with Additional Features and Fine-tuning on Custom Dataset (Issue #3)



linyihang1215 left a comment (PengYueECUST1997/CLEF#3)
 
Hello, I would like to ask how to obtain indicators for evaluating model performance and how to test model performance. I look forward to your reply. Thank you.

**Comment by YihangLin123** [comment]
Thank you very much for your prompt response to my question. I have two more questions. I would be grateful if you can respond to them.

① Is there any code for testing the model in the project? How is the performance of the CLEF model combined with biometrics and T3, T4, and T6 classifiers tested?

② Based on the repository you provided, I trained the CLEF model for the four biometrics you proposed. When predicting the Demo data, I could not get the same effect as your model.

I hope to get your reply. Thank you!

![Image](https://github.com/user-attachments/assets/0d245074-7387-49d3-9e13-bc166fc545d2)

![Image](https://github.com/user-attachments/assets/29bb5090-7ed2-459d-a771-e3db59d9e497)

![Image](https://github.com/user-attachments/assets/98ec7fb9-6517-401f-aead-a14bbdfa682a)

**Comment by YihangLin123** [comment]
I would also like to ask you, how do you train T3, T4, and T6 classifiers? How are the above classifiers obtained?

![Image](https://github.com/user-attachments/assets/47f140b3-21f2-471b-b385-3c4648814b9f)

**Comment by PengYueECUST1997** [comment]
Thank you very much for your message.
 
① Regarding your first question: I apologize that we did not include the testing code for the downstream classifier network in the repository. However, since this part involves a standard binary classification model, you can easily implement it by using the cross-modal representations generated from the contrastive learning pre-trained encoder as input features for a downstream classifier.
 
In our case, the downstream classifier is a simple multi-layer perceptron (MLP). You can refer to the test_dnn() function in Module.py for training. We used BCELoss as the loss function, with Dataset S2 for both training and validation. For training parameters, you can try setting the learning rate to 0.000045–0.000075, number of epochs to 40–55, and batch size to 32. (When training classifiers for T3/T4/T6SE prediction tasks, remember to assign binary labels accordingly for the proteins.)
 
After training your classifier, you can compare its classification performance on the independent test set (Dataset S3) using your own pre-trained encoder’s representations versus those from ESM-2 under the same conditions.
 
② As for your second question: the issue you encountered is likely because the classifier weight file you used does not match the pre-trained encoder you’re currently using. Even if you selected the same feature configurations, due to differences in weight initialization and the inherent randomness in training, the pre-trained encoder weights won’t be exactly the same.
 
Therefore, you would need to retrain a downstream classifier using the cross-modal representations generated by your own pre-trained encoder to obtain consistent and correct classification results.





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PengYueECUST1997/CLEF"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年6月6日(星期五) 下午4:43
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PengYueECUST1997/CLEF] Guidance on Improving Contrastive Learning with Additional Features and Fine-tuning on Custom Dataset (Issue #3)



linyihang1215 left a comment (PengYueECUST1997/CLEF#3)
 
Thank you very much for your prompt response to my question. I have two more questions. I would be grateful if you can respond to them.
 
① Is there any code for testing the model in the project? How is the performance of the CLEF model combined with biometrics and T3, T4, and T6 classifiers tested?
 
② Based on the repository you provided, I trained the CLEF model for the four biometrics you proposed. When predicting the Demo data, I could not get the same effect as your model.
 
I hope to get your reply. Thank you!
 
image.png (view on web)
 
image.png (view on web)
 
image.png (view on web)

**Comment by YihangLin123** [comment]
Thank you very much for your prompt reply. Your response has been very helpful to me! 

Thank you again for your excellent work, thank you!
